Source/JavaScriptCore/wasm/WasmBBQJIT.cpp

```diff
@@ -3787,18 +3793,131 @@ class BBQJIT {
+        StorageType elementType = getArrayElementType(typeIndex);
 
-        Vector<Value, 8> arguments = {
-            instanceValue(),
-            Value::fromI32(typeIndex),
-            arrayref,
-            index,
-            value,
-        };
-        // FIXME: Emit this inline.
-        // https://bugs.webkit.org/show_bug.cgi?id=245405
-        emitCCall(operationWasmArraySet, arguments);
+        Location arrayLocation;
+        if (arrayref.isPinned())
+            arrayLocation = locationOf(arrayref);
+        else
+            arrayLocation = loadIfNecessary(arrayref);
+        m_jit.loadPtr(MacroAssembler::Address(arrayLocation.asGPR(), JSWebAssemblyArray::offsetOfPayload()), wasmScratchGPR);
+
+        Location valueLocation;
+        if (value.isConst() && value.isFloat()) {
+            ScratchScope<0, 1> scratches(*this);
+            valueLocation = Location::fromFPR(scratches.fpr(0));
+            // Materialize the constant to ensure constant blinding.
+            emitMoveConst(value, valueLocation);
+        } else if (value.isConst()) {
+            ScratchScope<1, 0> scratches(*this);
+            valueLocation = Location::fromGPR(scratches.gpr(0));
+            // Materialize the constant to ensure constant blinding.
+            emitMoveConst(value, valueLocation);
+        } else
+            valueLocation = loadIfNecessary(value);
+        ASSERT(valueLocation.isRegister());
+
+        if (index.isConst()) {
+            ScratchScope<1, 0> scratches(*this);
+            auto fieldAddress = MacroAssembler::Address(wasmScratchGPR, JSWebAssemblyArray::offsetOfElements(elementType) + elementType.elementSize() * index.asI32());
+
+            Location valueLocation;
+            if (value.isConst() && value.isFloat()) {
+                ScratchScope<0, 1> scratches(*this);
+                valueLocation = Location::fromFPR(scratches.fpr(0));
+                // Materialize the constant to ensure constant blinding.
+                emitMoveConst(value, valueLocation);
+            } else if (value.isConst()) {
+                ScratchScope<1, 0> scratches(*this);
+                valueLocation = Location::fromGPR(scratches.gpr(0));
+                // Materialize the constant to ensure constant blinding.
+                emitMoveConst(value, valueLocation);
+            } else
+                valueLocation = loadIfNecessary(value);
+            ASSERT(valueLocation.isRegister());
+
+            if (elementType.is<PackedType>()) {
+                switch (elementType.as<Wasm::PackedType>()) {
+                case Wasm::PackedType::I8:
+                    m_jit.store8(valueLocation.asGPR(), fieldAddress);
+                    break;
+                case Wasm::PackedType::I16:
+                    m_jit.store16(valueLocation.asGPR(), fieldAddress);
+                    break;
+                }
+            } else {
+                ASSERT(elementType.is<Type>());
+                switch (value.type()) {
```

Comment:
What happens to ref types here? I don't see a write barrier

Reply:
Oh thanks for catching this, it seems to be missing. I had to refactor this code a few times so it may have gotten lost then. The write barrier code can be identical to the `struct.set` case.